source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/lac/petsc_communication_pattern.h>
+
+#ifdef DEAL_II_WITH_PETSC
+
+#  include <deal.II/base/mpi.h>
+
+#  include <deal.II/lac/exceptions.h>
+
+DEAL_II_NAMESPACE_OPEN
+// Shorthand notation for PETSc error codes.
+#  define AssertPETSc(code)                              \
+    do                                                   \
+      {                                                  \
+        PetscErrorCode __ierr = (code);                  \
+        AssertThrow(__ierr == 0, ExcPETScError(__ierr)); \
+      }                                                  \
+    while (0)
+
+namespace PETScWrappers
+{
+  CommunicationPattern::CommunicationPattern()
+    : sf(nullptr)
+  {}
+
+  CommunicationPattern::~CommunicationPattern()
+  {
+    clear();
+  }
+
+  void
+  CommunicationPattern::reinit(const types::global_dof_index local_size,
+                               const IndexSet &              ghost_indices,
+                               const MPI_Comm &              communicator)
+  {
+    clear();
+
+    PetscLayout layout;
+    AssertPETSc(PetscLayoutCreate(communicator, &layout));
+    AssertPETSc(PetscLayoutSetLocalSize(layout, local_size));
+    AssertPETSc(PetscLayoutSetUp(layout));
+
+    PetscInt start, end;
+    AssertPETSc(PetscLayoutGetRange(layout, &start, &end));
+
+    IndexSet want;
+    want.add_range(start, end);
+    want.add_indices(ghost_indices);
+    want.compress();
+
+    const PetscInt *idxs;
+    PetscInt        n;
+    IS              is = want.make_petsc_is(communicator);
+    AssertPETSc(ISGetLocalSize(is, &n));
```

Comment:
I don't understand this code. The `IS` object so created just gets a bunch of indices on every process, but no indication which of these are locally owned. How does it determine which ones of these are local?

Reply:
`PetscSFSetGraphLayout(sf, layout, n, nullptr, PETSC_OWN_POINTER, idxs))` does this for you.

Reply:
Using `IS` is superfluous here (I could have used a `std::vector<PetscInt>` instead), it is only to test the new functionality in `IndexSet`

Reply:
Still, I don't know what is happening in the three lines 66-68. You create an index set from a bunch of indices, and then `ISGetLocalSize` returns how many of these are stored on the current process: https://petsc.org/main/docs/manualpages/IS/ISGetLocalSize/ How does it know? Does `make_petsc_is()` internally partition indices into locally owned and ghost ones?

Reply:
the local size corresponds to the number of indices stored in the array, independent of their actual distribution.
`IS` does not know anything about global indices distribution, it is merely a class around an array of indices.

Reply:
So the indices returned are passed as "global indices I would like to receive" to the PetscSFSetGraphLayout call